frontend/dockerfile/dockerfile_test.go

```diff
@@ -6393,6 +6403,81 @@ ARG BUILDKIT_SBOM_SCAN_STAGE=true
 	require.Equal(t, 1, len(att.LayersRaw))
 }
 
+func testReproSourceDateEpoch(t *testing.T, sb integration.Sandbox) {
+	integration.CheckFeatureCompat(t, sb, integration.FeatureOCIExporter, integration.FeatureSourceDateEpoch)
+	f := getFrontend(t, sb)
+
+	tm := time.Date(2015, time.October, 21, 7, 28, 0, 0, time.UTC)
+	t.Logf("SOURCE_DATE_EPOCH=%d", tm.Unix())
+
+	dockerfile := []byte(`# The base image cannot be busybox, due to https://github.com/moby/buildkit/issues/3455
+FROM --platform=linux/amd64 debian:bullseye-20221219-slim@sha256:171530d298096f0697da36b3324182e872db77c66452b85783ea893680cc1b62
+RUN touch /foo
+RUN touch /foo.1
+RUN touch -d '2010-01-01 12:34:56' /foo-2010
+RUN touch -d '2010-01-01 12:34:56' /foo-2010.1
+RUN touch -d '2030-01-01 12:34:56' /foo-2030
+RUN touch -d '2030-01-01 12:34:56' /foo-2030.1
+RUN rm -f /foo.1
+RUN rm -f /foo-2010.1
+RUN rm -f /foo-2030.1
+
+# Limit the timestamp upper bound to SOURCE_DATE_EPOCH.
+# Workaround for https://github.com/moby/buildkit/issues/3180
+ARG SOURCE_DATE_EPOCH
+RUN find $( ls / | grep -E -v "^(dev|mnt|proc|sys)$" ) -newermt "@${SOURCE_DATE_EPOCH}" -writable -xdev | xargs touch --date="@${SOURCE_DATE_EPOCH}" --no-dereference
+
+# Squash the entire stage for resetting the whiteout timestamps.
+# Workaround for https://github.com/moby/buildkit/issues/3168
+FROM scratch
+COPY --from=0 / /
+`)
+
+	dir, err := integration.Tmpdir(
+		t,
+		fstest.CreateFile("Dockerfile", dockerfile, 0600),
+	)
+	require.NoError(t, err)
+	defer os.RemoveAll(dir)
+
+	c, err := client.New(sb.Context(), sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	outDigester := digest.SHA256.Digester()
+	outW := &iohelper.NopWriteCloser{Writer: outDigester.Hash()}
+
+	_, err = f.Solve(sb.Context(), c, client.SolveOpt{
+		FrontendAttrs: map[string]string{
+			"build-arg:SOURCE_DATE_EPOCH": fmt.Sprintf("%d", tm.Unix()),
+			"platform":                    "linux/amd64",
+		},
+		LocalDirs: map[string]string{
+			builder.DefaultLocalNameDockerfile: dir,
+			builder.DefaultLocalNameContext:    dir,
+		},
+		Exports: []client.ExportEntry{
+			{
+				Type: client.ExporterOCI,
+				Attrs: map[string]string{
+					// Remove buildinfo, as it contains the digest of the frontend image
+					"buildinfo": "false",
+				},
+				Output: fixedWriteCloser(outW),
+			},
+		},
+	}, nil)
+	require.NoError(t, err)
+
+	outDigest := outDigester.Digest().String()
+	t.Logf("OCI archive digest=%q", outDigest)
+	t.Log("The digest should be reproducible regardless to the kernel version, snapshotter, backing FS, etc.")
+	t.Log("However, the digest may divert across different versions of BuildKit.")
+
+	const expectedDigest = "sha256:d9128a57ccf5c137bd0168c71940ffdda344a0377953d0c496740bcd9cda9b9f"
+	require.Equal(t, expectedDigest, outDigest)
```

Comment:
Looks like the digest depends on the snapshotter.

The diff differs with the `native` snapshotter

```
--- overlayfs/blobs/sha256/7a0cfb79de7713bc3c5376b5f2dccf734f2466b0b5a3ea0950e009b465e9f200
+++ native/blobs/sha256/0b0bf9d38ac44add32536da2d02a3e9b73ae18ef0bdeca0fd813d865ff4c4bf7
│   --- 7a0cfb79de7713bc3c5376b5f2dccf734f2466b0b5a3ea0950e009b465e9f200-content
├── +++ 0b0bf9d38ac44add32536da2d02a3e9b73ae18ef0bdeca0fd813d865ff4c4bf7-content
│ ├── file list
│ │ @@ -50,15 +50,15 @@
│ │  -rwxr-xr-x   0 root         (0) root         (0)    39744 2015-10-21 07:28:00.000000 bin/sync
│ │  -rwxr-xr-x   0 root         (0) root         (0)   531928 2015-10-21 07:28:00.000000 bin/tar
│ │  -rwxr-xr-x   0 root         (0) root         (0)    14456 2015-10-21 07:28:00.000000 bin/tempfile
│ │  -rwxr-xr-x   0 root         (0) root         (0)   101408 2015-10-21 07:28:00.000000 bin/touch
│ │  -rwxr-xr-x   0 root         (0) root         (0)    39680 2015-10-21 07:28:00.000000 bin/true
│ │  -rwsr-xr-x   0 root         (0) root         (0)    35040 2015-10-21 07:28:00.000000 bin/umount
│ │  -rwxr-xr-x   0 root         (0) root         (0)    39744 2015-10-21 07:28:00.000000 bin/uname
│ │ --rwxr-xr-x   0 root         (0) root         (0)     2346 2015-10-21 07:28:00.000000 bin/uncompress
│ │ +hrwxr-xr-x   0 root         (0) root         (0)        0 2015-10-21 07:28:00.000000 bin/uncompress
│ │  -rwxr-xr-x   0 root         (0) root         (0)   147176 2015-10-21 07:28:00.000000 bin/vdir
│ │  -rwxr-xr-x   0 root         (0) root         (0)    63744 2015-10-21 07:28:00.000000 bin/wdctl
│ │  lrwxrwxrwx   0 root         (0) root         (0)        0 2015-10-21 07:28:00.000000 bin/ypdomainname -> hostname
│ │  -rwxr-xr-x   0 root         (0) root         (0)     1984 2015-10-21 07:28:00.000000 bin/zcat
│ │  -rwxr-xr-x   0 root         (0) root         (0)     1678 2015-10-21 07:28:00.000000 bin/zcmp
│ │  -rwxr-xr-x   0 root         (0) root         (0)     5898 2015-10-21 07:28:00.000000 bin/zdiff
│ │  -rwxr-xr-x   0 root         (0) root         (0)       29 2015-10-21 07:28:00.000000 bin/zegrep
│ │ @@ -654,15 +654,15 @@
│ │  -rwxr-xr-x   0 root         (0) root         (0)    72640 2015-10-21 07:28:00.000000 usr/bin/od
│ │  lrwxrwxrwx   0 root         (0) root         (0)        0 2015-10-21 07:28:00.000000 usr/bin/pager -> /etc/alternatives/pager
│ │  -rwxr-xr-x   0 root         (0) root         (0)   121072 2015-10-21 07:28:00.000000 usr/bin/partx
│ │  -rwsr-xr-x   0 root         (0) root         (0)    63960 2015-10-21 07:28:00.000000 usr/bin/passwd
│ │  -rwxr-xr-x   0 root         (0) root         (0)    43872 2015-10-21 07:28:00.000000 usr/bin/paste
│ │  -rwxr-xr-x   0 root         (0) root         (0)    39712 2015-10-21 07:28:00.000000 usr/bin/pathchk
│ │  -rwxr-xr-x   0 root         (0) root         (0)  3681152 2015-10-21 07:28:00.000000 usr/bin/perl
│ │ --rwxr-xr-x   0 root         (0) root         (0)  3681152 2015-10-21 07:28:00.000000 usr/bin/perl5.32.1
│ │ +hrwxr-xr-x   0 root         (0) root         (0)        0 2015-10-21 07:28:00.000000 usr/bin/perl5.32.1
│ │  -rwxr-xr-x   0 root         (0) root         (0)    44000 2015-10-21 07:28:00.000000 usr/bin/pinky
│ │  -rwxr-xr-x   0 root         (0) root         (0)    23112 2015-10-21 07:28:00.000000 usr/bin/pldd
│ │  -rwxr-xr-x   0 root         (0) root         (0)    76896 2015-10-21 07:28:00.000000 usr/bin/pr
│ │  -rwxr-xr-x   0 root         (0) root         (0)    39680 2015-10-21 07:28:00.000000 usr/bin/printenv
│ │  -rwxr-xr-x   0 root         (0) root         (0)    60256 2015-10-21 07:28:00.000000 usr/bin/printf
│ │  -rwxr-xr-x   0 root         (0) root         (0)    39672 2015-10-21 07:28:00.000000 usr/bin/prlimit
│ │  -rwxr-xr-x   0 root         (0) root         (0)   138336 2015-10-21 07:28:00.000000 usr/bin/ptx
```

Reply:
🤔 which one is correct?

Does it depend on snapshotter or differ?

Reply:
Ideally it should not depend, but it currently depends.

And the `native` implementation seems (unexpectedly?) more efficient than the `overlayfs` implementation.

Reply:
Are you saying that the optimized overlay differ loses some of the hardlink info? cc @ktock

Reply:
Seems so